modules/navigation/3d/nav_mesh_queries_3d.cpp

```diff
@@ -295,6 +314,40 @@ void NavMeshQueries3D::_query_task_build_path_corridor(NavMeshPathQueryTask3D &p
 
 				// Only consider the connection to another polygon if this polygon is in a region with compatible layers.
 				const NavBaseIteration *owner = connection.polygon->owner;
+				bool skip_connection = false;
+				if (p_query_task.exclude_regions || p_query_task.include_regions) {
+					switch (owner->get_type()) {
+						case NavigationUtilities::PathSegmentType::PATH_SEGMENT_TYPE_REGION: {
+							if (p_query_task.exclude_regions && p_query_task.excluded_regions.has(owner->get_self())) {
+								skip_connection = true;
+							}
+							if (p_query_task.include_regions && !p_query_task.included_regions.has(owner->get_self())) {
+								skip_connection = true;
+							}
+						} break;
+						case NavigationUtilities::PathSegmentType::PATH_SEGMENT_TYPE_LINK: {
+							const LocalVector<gd::Polygon> &link_polygons = owner->get_navmesh_polygons();
+							if (link_polygons.size() != 2) {
+								// Whatever this is, it is not a valid connected link.
+								skip_connection = true;
+							} else {
+								const RID link_start_region = link_polygons[0].owner->get_self();
+								const RID link_end_region = link_polygons[1].owner->get_self();
+								if (p_query_task.exclude_regions && p_query_task.excluded_regions.has(link_start_region)) {
```

Comment:
Don't you mean `link_end_region` here ? We should already be in the link_start_region,  no ? And can't nav link go both ways sometimes ? Don't we need to check both start and end if this is the case ?